internal/indexer/symbols.go

```diff
@@ -0,0 +1,196 @@
+package indexer
+
+import (
+	"fmt"
+	"go/ast"
+	"go/doc"
+	"go/token"
+	"log"
+	"os"
+	"strings"
+
+	protocol "github.com/sourcegraph/lsif-protocol"
+	"golang.org/x/tools/go/packages"
+)
+
+func (i *Indexer) indexSymbols() {
+	i.visitEachPackage("Indexing symbols", i.indexSymbolsForPackage)
+}
+
+// indexSymbolsForPackage emits the LSIF data associated with textDocument/documentSymbol for a
+// single Go package. See
+// https://microsoft.github.io/language-server-protocol/specifications/lsif/0.5.0/specification/#documentSymbol.
+//
+// This data includes:
+// - the textDocument/documentSymbol edge
+// - the documentSymbolResult vertex, which contains a hierarchy of RangeBasedDocumentSymbol vertices
+// - ranges associated with each symbol, with the appropriate range tag
+//
+// Because the textDocument/documentSymbol spec is document-based, all symbol hierarchies are rooted
+// at documents. Top-level variables, functions, types, constants, etc. are children of the root,
+// and methods are children of their receiver type.
+func (i *Indexer) indexSymbolsForPackage(p *packages.Package) {
+	documentSymbols := make(map[string]*symbolData)
+
+	// Omit files (such as those generated by `go test`) that aren't in the project root because
+	// those are not externally accessible.
+	files := make([]*ast.File, 0, len(p.Syntax))
+	for _, file := range p.Syntax {
+		path := p.Fset.Position(file.Pos()).Filename
+		if strings.HasPrefix(path, i.projectRoot) {
+			files = append(files, file)
+		}
+	}
+
+	// Protect with mutex, because multiple packages with the same package path can share files and
+	// doc.NewFromFiles will modify the elements of the file slice.
+	i.stripedMutex.LockKey(p.PkgPath)
+	docpkg, err := doc.NewFromFiles(p.Fset, files, p.PkgPath, doc.AllDecls|doc.PreserveAST)
+	if err != nil {
+		i.stripedMutex.UnlockKey(p.PkgPath)
+		fmt.Fprintf(os.Stderr, "Package %s aborted: doc.NewFromFiles failed with error: %s", p.PkgPath, err)
+		return
+	}
+	i.stripedMutex.UnlockKey(p.PkgPath)
```

Comment:
You need to do it manually instead of using a defer in order to (1) ensure it's unlocked and (2) keep the critical section guarded by the lock small. You'll see this idiom frequently around locks where one exit is a conditional early-exit.